plugins/modules/system/keyring.py

```diff
@@ -0,0 +1,282 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2022, Alexander Hussey <ahussey@redhat.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+"""
+Ansible Module - community.general.keyring
+"""
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+DOCUMENTATION = r"""
+---
+module: keyring
+version_added: 5.1.0
+author:
+  - Alexander Hussey (@ahussey-redhat)
+short_description: Set or delete a passphrase using the Operating System's native keyring
+description: >-
+  This module uses the L(keyring Python library, https://pypi.org/project/keyring/)
+  to set or delete passphrases for a given service and username from the OS' native keyring.
+requirements:
+  - keyring (Python library)
+  - gnome-keyring (application - required for headless Gnome keyring access)
+  - dbus-run-session (application - required for headless Gnome keyring access)
+options:
+  service:
+    description: The name of the service.
+    required: true
+    type: str
+  username:
+    description: The user belonging to the service.
+    required: true
+    type: str
+  user_password:
+    description: The password to set.
+    required: false
+    type: str
+    aliases:
+      - password
+  keyring_password:
+    description: Password to unlock keyring.
+    required: true
+    type: str
+  state:
+    description: Whether the password should exist.
+    required: false
+    default: present
+    type: str
+    choices:
+      - present
+      - absent
+"""
+
+EXAMPLES = r"""
+- name: test keyring module
+  hosts: localhost
+  connection: local
+  become: false
+  gather_facts: true
+  vars:
+    keyring_password: test
+    user_password: test12345
+  tasks:
+    - name: Set a password for test/test1
+      community.general.keyring:
+        service: test
+        username: test1
+        user_password: "{{ user_password }}"
+        keyring_password: "{{ keyring_password }}"
+      register: set_password
+      no_log: true
+
+    - name: Delete the password for test/test1
+      community.general.keyring:
+        service: test
+        username: test1
+        user_password: "{{ user_password }}"
+        keyring_password: "{{ keyring_password }}"
+        state: absent
+      register: delete_password
+      no_log: true
+"""
+
+try:
+    from shlex import quote
+except ImportError:
+    from pipes import quote
+from subprocess import Popen, PIPE
+import traceback
+
+from ansible.module_utils.basic import AnsibleModule, missing_required_lib
+
+try:
+    import keyring
+
+    HAS_KEYRING = True
+except ImportError:
+    HAS_KEYRING = False
+    KEYRING_IMP_ERR = traceback.format_exc()
+
+
+def del_passphrase(module):
+    """
+    Attempt to delete passphrase in the keyring using a shell rather than the Python API.
```

Comment:
I don't think this docstring is correct (and neither are the ones for the following functions). The code tries to use the Python API first, and falls back to using a shell if that doesn't work.

Reply:
That is correct. In early iteration of this module, I had these functions purely handle the alternative methods, but now the function tries the Python API method and falls back to the shell method